source/common/router/router.h

```diff
@@ -161,8 +163,20 @@ class Filter : Logger::Loggable<Logger::Id::router>,
     }
     return {};
   }
-  const Router::MetadataMatchCriteria* metadataMatchCriteria() const override {
+  const Router::MetadataMatchCriteria* metadataMatchCriteria() override {
```

Comment:
In the retry case you may end up calling this method multiple times. Is it worth checking to see if `metadata_match_` is not nullptr and returning if not? Is there any chance that doing the merge multiple times could have strange effects? If not to both of these, I would still add a comment about how this can happen multiple times for a single request.

Reply:
I think checking for nullptr and returning early is worth it. Multiple calls to merge should yield the same result, but we can avoid the computation and creating new objects.

Reply:
@mattklein123 hold on -- can the `route_entry_` for a request change (e.g.: someone called `clearRouteCache()` and then `metadataMatchCriteria()` gets called again)? In that case, caching the result for `metadataMatchCriteria()` would be wrong... we'd want to recompute.

cc: @zuercher @htuch

Reply:
I think it's ok, as long as metadataMatchCriteria() isn't invoked until LoadBalancer::chooseHost, which happens during Router::Filter::getConnPool in Filter::decodeHeaders and again in Filter::doRetry. The route_ and route_entry_ get set in decodeHeaders before load balancing, and I believe that once that happens they do not change.

clearRouteCache operates on the Http::ConnectionManagerImpl::ActiveStreamFilterBase which ends up being the source of the route_ and therefore route_entry_ via its implementation of StreamFilterCallbacks::route().